### PR TITLE
Obc setup plus segment update period

### PR DESCRIPTION
**Pull request by nikizadehgfdl**
This is two sets of updates for OBC in COBALT/OBGC tracers combined.

Setup OBC segments for COBALT/OBGC tracers
    
        - These are updates required to setup OBC segments for OBGC tracers.
        - Since COBALT package has more than 50 tracers using the MOM6 table
          mechanism for setting up OBC segments is not feasible. Rather, this
          update delegates such setup to mechanims used in ocean_BGS tracers
          leaving MOM6 mechanism for native tracers intact.
        - Fixed issues caught by MOM6 githubCI

Add capability to change obc segment update period
    
    - COBALT tracers do not need as frequent segment bc updates and can
      use a larger update period to speed up the model.
      This commit introduces a new parameter DT_OBC_SEG_UPDATE_OBGC
      that can be adjusted for obc segment update period.
    - This commit applies the change only to BGC tracers but can easily
      be changed to apply for all.

### CODE REVIEW

**Comment by kshedstrom** [comment]
The Bering domain with all the scale factors failed for me in real_to_time(CS%dt_obc_seg_period) where the dt variable was 1.3e15 instead of 600. This is in MOM.F90#L2895.

**Comment by nikizadehgfdl** [comment]
@kshedstrom did you set CS%dt_obc_seg_period to something? Where does 1.3e15 come from? How can I repeat your test? Thanks.

**Comment by kshedstrom** [comment]
Hi Niki, the files for the Bering domain are on gaea in /lustre/f2/dev/Katherine.Hedstrom/ESMG/ESMG-configs/Bering and the INPUT subdirectory. Check the MOM_override file for all the scalings I was testing. It's fine if you turn off all the scalings. Sorry, we're traveling, so I might not be able to get back to you right away if you need more from me.

File: src/core/MOM.F90 [hunk 1]
```diff
@@ -2856,6 +2880,12 @@ subroutine initialize_MOM(Time, Time_init, param_file, dirs, CS, restart_CSp, &
             CS%ntrunc, cont_stencil=CS%cont_stencil)
   endif
 
+  !Set OBC segment data update period
+  if (associated(CS%OBC) .and. CS%dt_obc_seg_period > 0.0) then
+    CS%dt_obc_seg_interval = real_to_time(CS%dt_obc_seg_period)
+    CS%dt_obc_seg_time = Time + CS%dt_obc_seg_interval
```
Comment (andrew-c-ross) [inline src/core/MOM.F90:2886]: I need to test this, but I think setting this to `Time + CS%dt_obc_seg_interval` results in the tracer reservoirs not being initialized at the first time step and only getting initialized at the first obc segment update period. In both NWA and NEP regional domains, water inflowing from the boundary is very low in alkalinity, oxygen, and other tracers when the model first starts running, as if the reservoirs were initialized at 0.

Comment (andrew-c-ross) [reply src/core/MOM.F90:2886]: I guess the reservoirs should also be initialized by `fill_obgc_segments`. Need to look into this more.

File: src/tracer/MOM_generic_tracer.F90 [hunk 2]
```diff
@@ -348,6 +398,8 @@ subroutine initialize_MOM_generic_tracer(restart, day, G, GV, US, h, param_file,
         call set_initialized(tr_ptr, g_tracer_name, CS%restart_CSp)
       endif
 
+      call g_tracer_get_obc_segment_props(g_tracer,g_tracer_name,obc_has )
+      if(obc_has .and. g_tracer_is_prog(g_tracer)) call fill_obgc_segments(G, GV, CS%OBC, tr_ptr, g_tracer_name)
```
Comment (andrew-c-ross) [inline src/tracer/MOM_generic_tracer.F90:402]: Replace `CS%OBC` with OBC. `CS%OBC` is the unneeded generic tracer OBC and is null.

File: src/core/MOM_open_boundary.F90 [hunk 3]
```diff
@@ -4648,6 +4744,119 @@ subroutine register_temp_salt_segments(GV, US, OBC, tr_Reg, param_file)
...
+        if (segment%direction == OBC_DIRECTION_S) then
+          segment%tr_Reg%Tr(nt)%t(i,J,k) = tr_ptr(i,j+1,k)
+        else
+          segment%tr_Reg%Tr(nt)%t(i,J,k) = tr_ptr(i,j,k)
+        endif
+      enddo ; enddo
+    endif
+    segment%tr_Reg%Tr(nt)%tres(:,:,:) = segment%tr_Reg%Tr(nt)%t(:,:,:)
+  enddo
+!  call setup_OBC_tracer_reservoirs(GV, OBC, .true.) ! Skip T and S to avoid restart bug
```
Comment (andrew-c-ross) [inline src/core/MOM_open_boundary.F90:4857]: This commented line is a leftover from old debugging and can be deleted. But, there is a necessary addition to this subroutine. The generic tracer segment data was not set up yet during MOM_initialize_state() -> fill_temp_salt_segments() -> setup_obc_tracer_reservoirs(), leaving OBC%tres_* 0 and causing low values to propagate in from the boundary initially. One way to fix this is to copy to OBC%tres_* again in this subroutine. I tried this [here](https://github.com/andrew-c-ross/MOM6/commit/c1acb365eb61f52e23f5bdd6c22758b0a909f192) and it works.

**Comment by nikizadehgfdl** [comment]
I made the following updates to this PR
- Add US%T_to_s unit conversions to dt that was causing @kshedstrom tests to crash
- Added @andrew-c-ross update to avoid low initial values for tracer reservoirs
- Merged in dev/gfdl

**Comment by andrew-c-ross** [comment]
Thanks Niki. I'm still getting tracer reservoirs that are initialized with zeros. I think we still need to add my first comment, replacing `CS%OBC` with just `OBC` in MOM_generic_tracer.F90. Do you agree that this is the correct OBC to use?

File: src/core/MOM.F90 [hunk 4]
```diff
@@ -1131,6 +1135,17 @@ subroutine step_MOM_dynamics(forces, p_surf_begin, p_surf_end, dt, dt_thermo, &
     call disable_averaging(CS%diag)
   endif
 
+  !OBC segment data update for some fields can be less frequent than others
```
Comment (Hallberg-NOAA) [inline src/core/MOM.F90:1138]: The intentation and spacing in this block of code do not comply with MOM6 standards, as articulated at https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide.

File: src/core/MOM_open_boundary.F90 [hunk 5]
```diff
@@ -346,6 +357,18 @@ module MOM_open_boundary
...
+!> Type to carry OBC information needed for setting segments for OBGC tracers
+type, private :: external_tracers_segments_props
+   type(external_tracers_segments_props), pointer :: next => NULL() !< pointer to the next node
+   character(len=128) :: tracer_name      !< tracer name
+   character(len=128) :: tracer_src_file  !< tracer source file for BC
+   character(len=128) :: tracer_src_field !< name of the field in source file to extract BC
+   real               :: lfac_in  !< multiplicative factor for inbound  tracer reservoir length scale
+   real               :: lfac_out !< multiplicative factor for outbound tracer reservoir length scale
+end type external_tracers_segments_props
+!> Keeps the OBC segment properties for external tracers
```
Comment (Hallberg-NOAA) [inline src/core/MOM_open_boundary.F90:369]: As described at https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide#global--module-data, MOM6 does not permit the use of module data, apart from CPU clocks, apart from a few specific exceptions that do not apply to the new variables being introduced here.  Please refactor this contribution in alignment with this policy.

File: src/core/MOM_open_boundary.F90 [hunk 6]
```diff
@@ -762,9 +787,22 @@ subroutine initialize_segment_data(G, GV, US, OBC, PF)
     IsdB = segment%HI%IsdB ; IedB = segment%HI%IedB
     JsdB = segment%HI%JsdB ; JedB = segment%HI%JedB
 
-    do m=1,num_fields
-      call parse_segment_data_str(trim(segstr), m, trim(fields(m)), &
-          value, filename, fieldname)
+    obgc_segments_props_list => obgc_segments_props !Get a pointer to the saved type
```
Comment (Hallberg-NOAA) [inline src/core/MOM_open_boundary.F90:790]: This block of code uses module variables and it ignores the guidance in the MOM6 style guide. Please correct the indenting and white space, and avoid using saves types.

File: src/core/MOM_open_boundary.F90 [hunk 7]
```diff
@@ -4648,6 +4744,123 @@ subroutine register_temp_salt_segments(GV, US, OBC, tr_Reg, param_file)
...
+  node_ptr%lfac_in  = lfac_in
+  node_ptr%lfac_out = lfac_out
+  !Reversed Linked List implementation! Make this new node to be the head of the list.
+  node_ptr%next => obgc_segments_props
+  obgc_segments_props => node_ptr
+  num_obgc_tracers = num_obgc_tracers+1
+end subroutine set_obgc_segments_props
+
+!> Get the OBC properties of external obgc tracers, such as their source file, field name,
+!  reservoir length scale factors
```
Comment (Hallberg-NOAA) [inline src/core/MOM_open_boundary.F90:4770]: The continuation line of this comment should start with `!!` to use the right doxygen syntax.

File: src/tracer/MOM_generic_tracer.F90 [hunk 8]
```diff
@@ -216,6 +220,52 @@ function register_MOM_generic_tracer(HI, GV, param_file, CS, tr_Reg, restart_CS)
...
+
+    if (.NOT. associated(OBC)) return
+    !Get the tracer list
+    call generic_tracer_get_list(CS%g_tracer_list)
+    if (.NOT. associated(CS%g_tracer_list)) call MOM_error(FATAL, trim(sub_name)//&
+         ": No tracer in the list.")
+
+    g_tracer=>CS%g_tracer_list
+    do
+      call g_tracer_get_alias(g_tracer,g_tracer_name)
```
Comment (Hallberg-NOAA) [inline src/tracer/MOM_generic_tracer.F90:251]: Please modify the following block to respect the MOM6 style guide with respect to indenting and the use of a space after an if, so that `if(obc_has)` will instead read `if (obc_has)`.

File: src/tracer/MOM_tracer_flow_control.F90 [hunk 9]
```diff
@@ -357,6 +359,26 @@ subroutine tracer_flow_control_init(restart, day, G, GV, US, h, param_file, diag
 
 end subroutine tracer_flow_control_init
 
+!> This subroutine calls all registered tracers to register their OBC segments
+!  similar to register_temp_salt_segments for T&S
```
Comment (Hallberg-NOAA) [inline src/tracer/MOM_tracer_flow_control.F90:363]: The continuation line of this doxygen comment should start with `!!`.

**Review by andrew-c-ross — APPROVED** [review]
This looks good to me, and I ran some short experiments with it with no apparent issues. Thanks for pushing this through, Niki.

**Review by Hallberg-NOAA — APPROVED** [review]
This updated PR has now addressed the concerns that I had raised with the previous version. It should be accepted once it has passed the pipeline testing.

File: src/core/MOM.F90 [hunk 10]
```diff
@@ -2152,6 +2167,13 @@ subroutine initialize_MOM(Time, Time_init, param_file, dirs, CS, restart_CSp, &
                  units="s", default=default_val, do_not_read=(dtbt > 0.0))
   endif
 
+  CS%dt_obc_seg_period = -1.0
+  call get_param(param_file, "MOM", "DT_OBC_SEG_UPDATE_OBGC", CS%dt_obc_seg_period, &
```
Comment (Hallberg-NOAA) [inline src/core/MOM.F90:2171]: As written, this new parameter is being logged in every MOM_parameter_doc.all file, including in many where it is meaningless.  Please add the argument `do_not_log=.not.associated(CS%OBC)` to this call or wrap it in on logical test for whether  open boundary conditions are actually being used.

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
This PR conditionally passed the pipeline testing at https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17365, but every MOM_parameter_doc.all file was changed, including those for which these change are not pertinent.

Please modify this PR as described in a separate comment on the relevant line to avoid adding the new runtime parameter DT_OBC_SEG_UPDATE_OBGC to the MOM_parameter_doc.all files when it is irrelevant.

**Review by Hallberg-NOAA — APPROVED** [review]

**Comment by Hallberg-NOAA** [comment]
This PR is undergoing a repeat of the pipeline testing at https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17391.

**Comment by Hallberg-NOAA** [comment]
Because of all of the merges and commits with minor corrections with this PR, I am going to handle this via a squash merge.
